pkg/sync/syncWorker.go

```diff
@@ -0,0 +1,205 @@
+package sync
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/xmtp/xmtpd/pkg/db/queries"
+	"github.com/xmtp/xmtpd/pkg/proto/xmtpv4/message_api"
+	"github.com/xmtp/xmtpd/pkg/registrant"
+	"github.com/xmtp/xmtpd/pkg/registry"
+	"github.com/xmtp/xmtpd/pkg/tracing"
+	"go.uber.org/zap"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+	"google.golang.org/protobuf/proto"
+)
+
+type syncWorker struct {
+	ctx          context.Context
+	log          *zap.Logger
+	nodeRegistry registry.NodeRegistry
+	registrant   *registrant.Registrant
+	store        *sql.DB
+	wg           sync.WaitGroup
+}
+
+func startSyncWorker(
+	ctx context.Context,
+	log *zap.Logger,
+	nodeRegistry registry.NodeRegistry,
+	registrant *registrant.Registrant,
+	store *sql.DB,
+) (*syncWorker, error) {
+	s := &syncWorker{
+		ctx:          ctx,
+		log:          log.With(zap.String("method", "syncWorker")),
+		nodeRegistry: nodeRegistry,
+		registrant:   registrant,
+		store:        store,
+		wg:           sync.WaitGroup{},
+	}
+	if err := s.start(); err != nil {
+		return nil, err
+	}
+	return s, nil
+}
+
+func (s *syncWorker) start() error {
+	nodes, err := s.nodeRegistry.GetNodes()
+	if err != nil {
+		return err
+	}
+	for _, node := range nodes {
+		if node.NodeID == s.registrant.NodeID() || !node.IsHealthy {
+			continue
+		}
+		s.subscribeToNode(node)
+	}
+	return nil
+}
+
+func (s *syncWorker) close() {
+	s.wg.Wait()
+}
+
+func (s *syncWorker) subscribeToNode(node registry.Node) {
+	tracing.GoPanicWrap(
+		s.ctx,
+		&s.wg,
+		fmt.Sprintf("node-subscribe-%d", node.NodeID),
+		func(ctx context.Context) {
+			var err error
+			for {
+				if err != nil {
+					log.Error(fmt.Sprintf("Error: %v, retrying...", err))
+					time.Sleep(1 * time.Second)
```

Comment:
this should eventually be done by an exponential backoff

Reply:
Agree, will work on that as part of handling node failures